[Monaco Editor] Add Search functionality

ElenaStoeva opened the pull request:
Fixes https://github.com/elastic/kibana/issues/186635

## Summary

This PR adds the find action to the Monaco code editor, which enables the search bar functionality as it is needed for Console with Monaco. The functionality is disabled by default.

<img width="1496" alt="Screenshot 2024-07-15 at 18 08 14" src="https://github.com/user-attachments/assets/a7476e4c-8ab8-4523-a815-1f7db483ddfa">

**How to test:**
1. Open Console and verify that Command/Ctrl + f opens the search bar (should work for both the editor and the output panel).
2. Open any other code editor in Kibana and verify that the search functionality is disabled.


<!--
### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
-->

ElenaStoeva committed:
[Monaco Editor] Add Search functionality

ElenaStoeva commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-management (Team:Kibana Management)

eokoneyo reviewed (changes requested):
I think we should make the find functionality opt-in as oppose to making it available by default to all consumers. For instance this change would show up in discover for ES|QL like so; 

<img width="2558" alt="Screenshot 2024-07-16 at 15 45 32" src="https://github.com/user-attachments/assets/bdec9458-18a4-4b0c-a1c2-b3ebd2045679">

There's a [monaco API](https://microsoft.github.io/monaco-editor/docs.html#functions/editor.addKeybindingRule.html) to modify key bindings that we can use to disable the find action if a user doesn't opt in.

*addendum*
See https://github.com/microsoft/monaco-editor/discussions/3784 for the usage of this API

ElenaStoeva committed:
Disable find action by default

ElenaStoeva commented:
Thanks for the review and for the suggestion @eokoneyo! I addressed your feedback in https://github.com/elastic/kibana/pull/188337/commits/3504ed1611326a898bb00d4036a0d5c24489cade.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/221814)
* Commit: 3504ed1611326a898bb00d4036a0d5c24489cade

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/c5165d9afe9b66c0b4b55bf1b62f7edea897cc44) | [after](https://github.com/elastic/kibana/commit/3504ed1611326a898bb00d4036a0d5c24489cade) | diff |
| --- | --- | --- | --- |
| `console` | 457.0KB | 457.1KB | +40.0B |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/c5165d9afe9b66c0b4b55bf1b62f7edea897cc44) | [after](https://github.com/elastic/kibana/commit/3504ed1611326a898bb00d4036a0d5c24489cade) | diff |
| --- | --- | --- | --- |
| `kbnUiSharedDeps-css` | 206.6KB | 211.9KB | +5.3KB |
| `kbnUiSharedDeps-srcJs` | 3.2MB | 3.2MB | +72.3KB |
| total |  |  | +77.6KB |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/c5165d9afe9b66c0b4b55bf1b62f7edea897cc44) | [after](https://github.com/elastic/kibana/commit/3504ed1611326a898bb00d4036a0d5c24489cade) | diff |
| --- | --- | --- | --- |
| `@kbn/code-editor` | 40 | 41 | +1 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/c5165d9afe9b66c0b4b55bf1b62f7edea897cc44) | [after](https://github.com/elastic/kibana/commit/3504ed1611326a898bb00d4036a0d5c24489cade) | diff |
| --- | --- | --- | --- |
| `@kbn/code-editor` | 2 | 3 | +1 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/c5165d9afe9b66c0b4b55bf1b62f7edea897cc44) | [after](https://github.com/elastic/kibana/commit/3504ed1611326a898bb00d4036a0d5c24489cade) | diff |
| --- | --- | --- | --- |
| `@kbn/code-editor` | 2 | 3 | +1 |


</details>


### History
* :green_heart: [Build #221542](https://buildkite.com/elastic/kibana-pull-request/builds/221542) succeeded 7f2a4bbe62f77e0c991fab9b5f993c108ac8ae01

cc @ElenaStoeva

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/221814","number":221814,"commit":"3504ed1611326a898bb00d4036a0d5c24489cade"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/221542","number":221542,"commit":"7f2a4bbe62f77e0c991fab9b5f993c108ac8ae01"}],"number":221814}
buildkite-pr-comment-->

eokoneyo commented on packages/shared-ux/code_editor/impl/code_editor.tsx:
```diff
@@ -188,6 +193,7 @@ export const CodeEditor: React.FC<CodeEditorProps> = ({
   }),
   fitToContent,
   accessibilityOverlayEnabled = true,
+  enableFindAction = false,
```
We don't actually need to specify false, undefined is also a falsy value.

eokoneyo reviewed (approved):
@ElenaStoeva tested this, it works perfectly well. Thanks for making the change

ElenaStoeva committed:
Remove default false value of prop

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/222010)
* Commit: bab9b40088d46b6f479d825d1dc188ad2d85d758

### Failed CI Steps
* [Jest Integration Tests #5](https://buildkite.com/elastic/kibana-pull-request/builds/222010#0190c03b-9650-4a6e-9096-0e6d479d0b71)
* [Jest Integration Tests #5](https://buildkite.com/elastic/kibana-pull-request/builds/222010#0190c06b-96b2-4a76-be05-13e2e9147b1a)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222010#0190c03b-9650-4a6e-9096-0e6d479d0b71) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222010/jobs/0190c03b-9650-4a6e-9096-0e6d479d0b71/artifacts/0190c06b-4aa5-4bea-b224-15b6c1095da9) Jest Integration Tests #<span></span>5 / telemetry tasks telemetry-prebuilt-rule-alerts should manage runtime errors searching endpoint metrics
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/222010#0190c06b-96b2-4a76-be05-13e2e9147b1a) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/222010/jobs/0190c06b-96b2-4a76-be05-13e2e9147b1a/artifacts/0190c087-ac83-4f21-b1b0-6858cb009e87) Jest Integration Tests #<span></span>5 / telemetry tasks telemetry-prebuilt-rule-alerts should manage runtime errors searching endpoint metrics

### History
* :green_heart: [Build #221814](https://buildkite.com/elastic/kibana-pull-request/builds/221814) succeeded 3504ed1611326a898bb00d4036a0d5c24489cade
* :green_heart: [Build #221542](https://buildkite.com/elastic/kibana-pull-request/builds/221542) succeeded 7f2a4bbe62f77e0c991fab9b5f993c108ac8ae01

cc @ElenaStoeva

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/222010","number":222010,"commit":"bab9b40088d46b6f479d825d1dc188ad2d85d758"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/221814","number":221814,"commit":"3504ed1611326a898bb00d4036a0d5c24489cade"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/221542","number":221542,"commit":"7f2a4bbe62f77e0c991fab9b5f993c108ac8ae01"}],"number":222010}
buildkite-pr-comment-->

ElenaStoeva committed:
Merge branch 'main' into monaco-editor/add-search-functionality

yuliacech commented on packages/shared-ux/code_editor/impl/code_editor.tsx:
```diff
@@ -375,6 +381,12 @@ export const CodeEditor: React.FC<CodeEditorProps> = ({
+      monaco.editor.addKeybindingRule({
+        // eslint-disable-next-line no-bitwise
+        keybinding: monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyF,
+        command: enableFindAction ? 'actions.find' : null,
```
nit: instead of registering the keybindings with a null action, I think we could only register the key binding if the find action is enabled

ElenaStoeva replied:
Since we added the find action import in `packages/kbn-monaco/src/monaco_imports.ts`, this action is now enabled by default in every Monaco editor, unless we disable it. Therefore, we need to specify explicitly that the key binding should run no action; otherwise it would run the find action by default.

yuliacech reviewed (approved):
Thanks a lot for adding the search bar, @ElenaStoeva! 
Tested locally and all works as expected 👍

ElenaStoeva committed:
Merge branch 'main' into monaco-editor/add-search-functionality
